test-loop-tests/src/tests/global_contracts_distribution.rs

```diff
@@ -0,0 +1,186 @@
+    // This height is picked so that the first global contract distribution receipt reaches
+    // shard that is being split at the first height after the resharding
+    let send_deploy_tx_height = expected_new_shard_layout_height - 3;
+
+    run_until_node_head_height(
+        &mut env.env,
+        &env.chunk_producer,
+        send_deploy_tx_height,
+        Duration::seconds(10),
+    );
+    assert_eq!(env.current_shard_layout(), env.base_shard_layout);
+
+    // Deploying global contract with the user from the split shard.
+    // The first target shard for the global contract distribution receipt
+    // is user's shard, this way we ensure that we hit the shard that is
+    // split at resharding.
+    let deploy_user = env.users[0].clone();
+    let code = ContractCode::new(near_test_contracts::rs_contract().to_vec(), None);
+    let deploy_tx = deploy_global_contract(
+        &mut env.env.test_loop,
+        &env.env.node_datas,
+        &env.chunk_producer,
+        deploy_user,
+        code.code().to_vec(),
+        1,
+        GlobalContractDeployMode::CodeHash,
+    );
+
+    run_until_node_head_height(
+        &mut env.env,
+        &env.chunk_producer,
+        expected_new_shard_layout_height,
+        Duration::seconds(10),
+    );
+    check_txs(&mut env.env.test_loop.data, &env.env.node_datas, &env.chunk_producer, &[deploy_tx]);
+    assert_eq!(env.current_shard_layout(), env.new_shard_layout);
+
+    // Verify that global contract distribution receipt has target shard from the old shard layout,
+    // while its block has the new layout.
+    {
+        let block =
+            env.client().chain.get_block_by_height(expected_new_shard_layout_height).unwrap();
+        let block_shard_layout = env
+            .client()
+            .epoch_manager
+            .get_epoch_config(block.header().epoch_id())
+            .unwrap()
+            .shard_layout;
+        assert_eq!(block_shard_layout, env.new_shard_layout);
+        let chunks = block.chunks();
+        let MaybeNew::New(chunk_header) = chunks.iter().nth(0).unwrap() else {
+            panic!("expected new chunk");
+        };
+        let chunk = env.client().chain.get_chunk(&chunk_header.compute_hash()).unwrap();
+        let distribution_receipt = chunk
+            .prev_outgoing_receipts()
+            .iter()
+            .filter_map(|r| match r.receipt() {
+                ReceiptEnum::GlobalContractDistribution(r) => Some(r),
+                _ => None,
+            })
+            .next()
+            .unwrap();
+        let target_shard = distribution_receipt.target_shard();
+        assert!(!block_shard_layout.shard_ids().contains(&target_shard));
+    };
+
+    // Wait for the distribution to reach all shards.
+    env.env.test_loop.run_for(Duration::seconds(3));
+
+    // Check that users on all shards in the new layout can use the contract.
+    let mut use_txs = vec![];
+    for user in &env.users {
```

Comment:
Does env have a user per shard in the new or old shard layout?

Reply:
new layout, let me add assert for that